datastore/mongodb/filter_outputs.go

```diff
@@ -31,47 +31,80 @@ func (c *Client) CreateFilterOutput(ctx context.Context, f *model.FilterOutput)
 func (c *Client) UpdateFilterOutput(ctx context.Context, f *model.FilterOutput) error {
 	col := c.collections.filterOutputs
 
-	var docs []model.FilterOutput
-	sc := bson.M{"id": f.ID}
-
-	var err error
-	var num int
-
-	if num, err = c.conn.Collection(col.name).Find(ctx, sc, &docs, mongodb.Limit(1)); err != nil {
-		return errors.Wrap(err, "failed to update filter output")
-	}
-	if num < 1 {
-		return &er{
-			err:      errors.Errorf("failed to find filter output"),
-			notFound: true,
-		}
+	docs, err := c.findFilterOutput(f.ID, ctx)
+	if err != nil {
+		return err
 	}
-
 	//a record with stat 'completed' can't be updated further
 	if docs[0].State == model.Completed {
-		return &er{
-			err:       errors.Errorf(`filter output is already in "completed" state`),
-			forbidden: true,
+		err := &er{
+			err: errors.Errorf(`filter output is already in "completed" state`),
 		}
+		err.forbidden = true
+		return err
 	}
 
-	updates := bson.M{"$set": bson.M{"state": f.State, "downloads": f.Downloads}}
+	nv := bson.M{"$set": bson.M{"state": f.State, "downloads": f.Downloads}}
 
 	var rec *mongodb.CollectionUpdateResult
+	sc := bson.M{"id": f.ID}
 
-	if rec, err = c.conn.Collection(col.name).Update(ctx, sc, updates); err != nil {
+	if rec, err = c.conn.Collection(col.name).Update(ctx, sc, nv); err != nil {
 		return errors.Wrap(err, "failed to update filter output")
 	}
 
 	//This should not happen unless the recored is being removed btween the first find and the update.
 	//Check if the update failed sue to search condition.
 	//Update returns no error but the result's MatchedCount has to be checked for number of updated records
 	if rec.MatchedCount < 1 {
+		err := &er{
+			err: errors.Errorf("failed to find filter output"),
+		}
+		err.notFound = true
+		return err
+	}
+
+	return nil
+}
+
+//AddFilterOutputEvent will append to the existing list of events in the filter output
+func (c *Client) AddFilterOutputEvent(ctx context.Context, id string, f *model.Event) error {
+	col := c.collections.filterOutputs
+
+	ne := map[string]model.Event{
+		"events": *f,
+	}
+
+	var rec *mongodb.CollectionUpdateResult
+	var err error
+	if rec, err = c.conn.Collection(col.name).Update(ctx, bson.M{"id": id}, bson.M{"$push": ne}); err != nil {
+		return errors.Wrap(err, "failed to add element to filter output")
+	}
+
+	if rec.MatchedCount != 1 {
 		return &er{
-			err:      errors.Errorf("failed to find filter output"),
+			err:      errors.Errorf("filter output not found"),
 			notFound: true,
 		}
 	}
 
 	return nil
 }
+
+func (c *Client) findFilterOutput(fId string, ctx context.Context) ([]model.FilterOutput, error) {
+	col := c.collections.filterOutputs
+	var docs []model.FilterOutput
+	sc := bson.M{"id": fId}
+	var err error
+	var num int
+	if num, err = c.conn.Collection(col.name).Find(ctx, sc, &docs, mongodb.Limit(1)); err != nil {
+		return nil, err
```

Comment:
We don't have a specific error to message for this case, so can't see the benefit of wrapping this. 
The wrap will look like something like this 
&er{
err : err
}

Reply:
I'm not sure what you mean by that? We want to wrap any error with a brief explanation of what's causing the error, this is so we understand precisely what the error is pointing to. Unless it will be redundant (i.e. no point returning `failed to find filter: failed to find filter`), we always want to add the extra information at each step so at the end we have complete description of what's gone wrong. Wrapping the error is about constructing the full error string, not just adding extra information like statusCodes and logData.

Reply:
At the moment an error might read something like `failed add filter output event: invalid query` but that doesn't explain the exact database call that's failing as you're making more than one, if you wrap at each level it would read either `failed add filter output event: failed to get filter output: invalid query` or `failed add filter output event: failed to add element to filter output: invalid query`, so you can see it helps us to differentiate what's failed